File: 300.longestIncreasingSubsequence/longestIncreasingSubsequence_nlogn.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums.length == 1) return 1;
        List<Integer> tails = new ArrayList<>();
        tails.add(nums[0]);
        for (int i = 1; i < nums.length; i++) {
            int index = findFirstGreaterOrEqualThan(tails, nums[i]);
            if (index == -1) {
                tails.add(nums[i]);
            } else {
                tails.set(index, nums[i]);
            }
        }
        return tails.size();
    }
    
    public int findFirstGreaterOrEqualThan(List<Integer> tails, int target) {
        int len = tails.size();
        int left = 0, right = len - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (tails.get(mid) <= target) {
                left = mid;
            } else {
                right = mid;
            }
        }
        if (tails.get(left) >= target) return left;
        else if (tails.get(right) >= target) return right;
        else return -1;
    }
}

File: 300.longestIncreasingSubsequence/longestIncreasingSubsequence_nlogn.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n == 0)
            return 0;
        vector<int> tail;
        for (int i = 0; i < n; i++) {
            int index = find_last_less_than(tail, nums[i]);
            if (index == tail.size())
                tail.push_back(nums[i]);
            else
                tail[index] = nums[i];
        }
        return tail.size();
    }
    int find_last_less_than(vector<int>& tail, int target) {
        int left = 0, right = tail.size();
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (tail[mid] < target)
                left = mid + 1;
            else
                right = mid;
        }
        return left;
    }
};

int main()
{
    Solution s;
    vector<int> nums = {10,9,2,5,3,7,101,18};
    cout << s.lengthOfLIS(nums) << endl;
    return 0;
}
